source/extensions/filters/udp/dns_filter/dns_parser.cc

```diff
@@ -210,20 +392,226 @@ DnsQueryRecordPtr DnsMessageParser::parseDnsQueryRecord(const Buffer::InstancePt
+  if (context->response_code_ == DNS_RESPONSE_CODE_FORMAT_ERROR) {
+    return;
+  }
+  // Check for unsupported request types
+  for (const auto& query : context->queries_) {
+    if (query->type_ != DNS_RECORD_TYPE_A && query->type_ != DNS_RECORD_TYPE_AAAA) {
+      context->response_code_ = DNS_RESPONSE_CODE_NOT_IMPLEMENTED;
+      return;
+    }
+  }
+  // Output validation
+  if (serialized_queries == 0) {
+    context->response_code_ = DNS_RESPONSE_CODE_FORMAT_ERROR;
+    return;
+  }
+  if (serialized_answers == 0) {
+    context->response_code_ = DNS_RESPONSE_CODE_NAME_ERROR;
+    return;
+  }
+  context->response_code_ = DNS_RESPONSE_CODE_NO_ERROR;
+}
+
+void DnsMessageParser::generateRandomIndices(const size_t count,
+                                             absl::flat_hash_set<size_t>& elements) {
+  elements.clear();
+  elements.reserve(count);
+  while (elements.size() < count) {
+    size_t element = rng_.random() % count;
+    if (!elements.contains(element)) {
+      elements.emplace(element);
+    }
+  }
+}
+
+void DnsMessageParser::buildResponseBuffer(DnsQueryContextPtr& query_context,
+                                           Buffer::OwnedImpl& buffer) {
+  // Ensure that responses stay below the 512 byte byte limit. If we are to exceed this we must add
+  // DNS extension fields
+  //
+  // Note:  There is Network::MAX_UDP_PACKET_SIZE, which is defined as 1500 bytes. If we support
+  // DNS extensions, which support up to 4096 bytes, we will have to keep this 1500 byte limit in
+  // mind.
+  static constexpr uint64_t MAX_DNS_RESPONSE_SIZE = 512;
+  static constexpr uint64_t MAX_DNS_NAME_SIZE = 255;
+
+  // Amazon Route53 will return up to 8 records in an answer
+  // https://aws.amazon.com/route53/faqs/#associate_multiple_ip_with_single_record
+  static constexpr size_t MAX_RETURNED_RECORDS = 8;
+
+  // Each response must have DNS flags, which spans 4 bytes. Account for them immediately so that we
+  // can adjust the number of returned answers to remain under the limit
+  uint64_t total_buffer_size = sizeof(DnsHeaderFlags);
+  uint16_t serialized_answers = 0;
+  uint16_t serialized_queries = 0;
+
+  Buffer::OwnedImpl query_buffer{};
+  Buffer::OwnedImpl answer_buffer{};
+
+  ENVOY_LOG(trace, "Building response for query ID [{}]", query_context->id_);
+
+  for (const auto& query : query_context->queries_) {
+    if (!query->serialize(query_buffer)) {
+      ENVOY_LOG(debug, "Unable to serialize query record for {}", query->name_);
+      continue;
+    }
+
+    // Serialize and account for each query's size. That said, there should be only one query.
+    ++serialized_queries;
+    total_buffer_size += query_buffer.length();
+
+    const auto& answers = query_context->answers_;
+
+    // Generate a set of up to 8 random indices of the answer records that we serialize. We do this
+    // so that we don't always return the same records in the same order for every query.
+    absl::flat_hash_set<size_t> indices{};
```

Comment:
In the case where they are less than 8 records, you can avoid the set entirely which I think is probably worth doing.

Also, I'm slightly worried about the random algorithm in the sense that it may require quite a lot of picks to finally converge. Would it be better to just randomize when you read the records in, and then if > max record size pick a random index to start? I think that would be more efficient and avoid the set in the runtime case? wdyt?

/wait-any

Reply:
I did find a way to use `std::shuffle` and `std::iota` so that we wouldn't have to use the set to dedupe entries.   That might eliminate the convergence issue since the index set is generated at once.  Sadly both functions are linear, but with a maximum of 8 entries, it might be acceptable.

Let me work on this and see what I can come up with.  I'm not sure that randomizing the input address list will not have similar concerns.